### ISSUE TITLE
Movie View

### ISSUE THREAD

**Issue by cdbzb** [supercollider/supercollider#4715]
## Motivation

For multimedia installations, and for composition to picture.

Some of this is possible now using external apps and OSC but these solutions leave much to be desired.

## Description of Proposed Feature

Allow Views to display video - it looks as though this existed in deprecated GUI kits in the past perhaps ?

**Comment by jamshark70** [comment]
> it looks as though this existed in deprecated GUI kits in the past perhaps?

The old Cocoa GUI scheme did have SCMovieView, and I believe SwingOSC might have had JSCMovieView (but I don't remember).

> Some of this is possible now using external apps and OSC but these solutions leave much to be desired.

It's worth expanding on that. 

One concrete drawback of any external solution is that you can't integrate it with SC GUIs (e.g. UserView).

To my knowledge, available solutions are:

- Pure Data + Gem (or Ofelia) objects [macOS, Win, Linux]: Pretty straightforward for basic cases, also supports rotation/scaling etc. (Win10 seems to require installing third-party, nonfree codecs, not guaranteed to work in all cases.)
  - Max + Jitter: Should be about the same, but no Linux.
- xjadeo [macOS, Win, Linux]: *Very* straightforward if all you want is a window showing the video frames.
- OSC control of a video-capable DAW: Fine, but heavy.

Pd/Gem capabilities make me think -- it would be nice to integrate this into UserView in some way. UserView already supports rotate/translate/scale operations; you might need some compositing options as well.

Probably this eventually needs to become a RFC. I don't think a QMovieView would be very complex, but using video as part of a more elaborate artistic display calls for other features, and the design would have to be discussed (e.g., would it be `Pen.drawVideoFrame(...)` or something else?).

**Comment by brianlheim** [comment]
https://doc.qt.io/qt-5/videooverview.html

**Comment by dyfer** [comment]
Adding QT video view would be nice! 

I'd just want to point that extra consideration could be put towards handling audio in video - _I don't think_ cocoa/quicktime video class provided that, but it's not unreasonable to expect the ability to manipulate audio of the video being played back (I'm thinking something like shared memory mechanism to transmit the audio to server's audio buses... but maybe that's too big of a complication)

**Comment by jamshark70** [comment]
> I'd just want to point that extra consideration could be put towards handling audio in video.

Is the expected model primarily to seek specific frames or to start/stop playback? I'd suggest the first as the primary case, with the second as a convenience. It may be feasible to run audio for the convenience method, but random seek, it's unclear what to do with the audio then. Many video editors play a short snip of audio for that frame, but it would be jarring in a musical context.

> Alternatively maybe an ability to load audio from video files into SoundFile/Buffer would be a solution, with some sync mechanism (sclang-based?) between audio and video.

Interesting. We've avoided directly handling compressed formats (at least ones not supported by libsndfile). One solution here would be to invoke `ffmpeg -i source.mp4 audio.wav` and then use the wav file... but the user may prefer to do this once and reuse the wav file rather than it being done every time, to a temp file.

The discrepancy between language and server clocks may become a practical issue, after that.

**Comment by cdbzb** [comment]
I'd imagine playback as the more common use case... ideally, one day, at
various speeds and with filtering...

I can't see the necessity of extracting audio as it seems easy enough to
handle offline.  Unless we are talking about live streaming or such...

**Comment by jamshark70** [comment]
> I'd imagine playback as the more common use case... ideally, one day, at various speeds and with filtering...

My reasoning is: If the implementation is based on seeking specific frames, then it's actually easy to implement playback on top of that. https://scsynth.org/t/display-video-or-video-frames/1570/8

But if the implementation is based on playback, it's harder to seek frames (without the ability to set speed to 0).

**Comment by telephon** [comment]
>I can't see the necessity of extracting audio as it seems easy enough to
handle offline. 

Using the audio from a movie is the most exceptional feature in this idea, just playing back a movie one can do with any external player via `systemCmd`.

**Comment by cdbzb** [comment]
@ Julian not disagreeing with your larger point but re external players,
for creative purposes you really want to be able to place and filter video
programmatically - draw over video certainly. A fantasy would be to access
rasterized video as a signal: for ancient video effects and to interact
with audio...

@James I am not super knowledgeable here so apologies if this is offbase
but my sense is that frame by frame rendering is computationally more
expensive and can have other problems vis playback - thinking
about compressed formats hardware acceleration etc. On the other hand it
could have advantages in terms of achieving sync. How does Max handle for
example.

**Comment by esluyter** [comment]
I'll share this in case it's relevant to the discussion... I spent years making a reliable (i.e. no visible or audible glitches or mid-show crashes ;) video playback system for a highly video based theater company, which turned out to be a _much_ more difficult task than anyone imagined. We needed the ability to play long (up to 4 hours!) videos (with precisely synced audio of course) forwards and backwards at arbitrary speeds (I think we've used 0.1x all the way up to 100x or so), to be able to smoothly ramp from one speed to another, and to be able to "scrub" i.e. seek individual frames.

We tried using all sorts of commercial and open software packages to achieve this, and literally all of the dedicated video softwares we tried were awful with audio rate scaling, and some of them crash when you jump around too much. We also tried Max/Jitter, which occasionally crashed as well and had some other deal breaking problems which I forget now.. The only reliable configuration we've found is to export the audio from each video file and load it into a SC buffer, then have SC drive video playback via position-seeking OSC (sending messages at 30 or 60 hz) to TouchDesigner.

Anyway, I've learned some lessons:
- I think frame seeking is absolutely necessary if you want to play at any sort of variable speed with uncompromised audio quality -- video renderers don't seem to be built for this sort of thing.
- video codec matters a lot when frame seeking, we've found HAP to be reliable when scrubbing forwards and backwards, more compressed formats like h.264 or 265 are ok forwards but not backwards (visible glitches)
- SC currently offers no out-of-the-box way to play long audio files and know precisely where you are in the file. BufRd is limited by 32-bit float precision to files that are just a few minutes long (and shorter if you want to play them slower); PlayBuf can play long files but you can't know where you are in them; and VDiskIn can notify you every time it reloads its buffer, which is I guess sometimes helpful but certainly not precise (also I think it has trouble playing backwards, and forwards at high speeds...). I wound up making a modded version of BufRd that takes two floats and adds them together to make a double-precision position, and a modded version of Phasor that outputs a double as two floats. (This is also a hack that was used in Max before it became 64 bit.)
- the only way I've found to smoothly handle playing audio files forward/backwards at greatly varying speeds is to load them entirely into RAM, which isn't ideal for probably most video playback scenarios. VDiskIn currently can't handle the task for several reasons mentioned above, and I remember it seemed like it would be a not-trivial task to modify it.

**Comment by jamshark70** [comment]
@esluyter absolutely, thanks for weighing in! It's easy to say "of course we ought to do x" but when in practice, x doesn't work as well as expected... your experiences here definitely should feed into any design decisions.

I'm certainly no expert on it -- all I've done is control either Gem [pix_film] or xjadeo by OSC from SuperCollider. Not a glitch-free result by any means.

You raise a good point about video/audio sync. Soundcard sample clocks are rarely accurate. Here in China, one of the streaming video sites doesn't account for this and audio drifts noticeably away from the video. I agree that it isn't a simple matter of running a DiskIn on the server and streaming video frames separately.

**Comment by muellmusik** [comment]
Just a random thought, as I have very little expertise in this, so apologies if this is very impractical. If we want to be able to process audio from video files and straightforwardly maintain sync (which seems very desirable to me!), would supporting video in the server be at least worth considering? (No idea about the technical challenges of that, tbf.) As VST support is already making it a GUI app perhaps that horse has left the barn?

**Comment by esluyter** [comment]
Interesting idea to play video on the server..... although my experience with video-related crashing and my bias towards audio stability makes me a little nervous about it.... also as I say the ongoing problem of sample-accurate seeking beyond the limits of 32 bit float precision
